// @screeps/engine/dist/game/tombstones.js
'use strict';

var utils = require('./../utils'),
    driver = utils.getRuntimeDriver(),
    C = driver.constants,
    _ = require('lodash');

var runtimeData, intents, register, globals;

function _storeGetter(o) {
    var result = { energy: o.energy };

    C.RESOURCES_ALL.forEach(function (resourceType) {
        if (o[resourceType]) {
            result[resourceType] = o[resourceType];
        }
    });

    return result;
}

exports.make = function (_runtimeData, _intents, _register, _globals) {

    runtimeData = _runtimeData;
    intents = _intents;
    register = _register;
    globals = _globals;

    if (globals.Tombstone) {
        return;
    }

    var data = function data(id) {
        if (!runtimeData.roomObjects[id]) {
            throw new Error("Could not find an object with ID " + id);
        }
        return runtimeData.roomObjects[id];
    };

    var Tombstone = register.wrapFn(function (id) {
        var _data = data(id);
        globals.RoomObject.call(this, _data.x, _data.y, _data.room);
        this.id = id;
    });

    Tombstone.prototype = Object.create(globals.RoomObject.prototype);
    Tombstone.prototype.constructor = Tombstone;
    utils.defineGameObjectProperties(Tombstone.prototype, data, {
        deathTime: function deathTime(o) {
            return o.deathTime;
        },
        store: _storeGetter,
        ticksToDecay: function ticksToDecay(o) {
            return o.decayTime - runtimeData.time;
        },
        creep: function creep(o) {
            var creep = new globals.Creep();
            globals.RoomObject.call(creep, o.x, o.y, o.room);
            Object.defineProperties(creep, {
                id: {
                    enumerable: true,
                    get: function get() {
                        return o.creepId;
                    }
                },
                name: {
                    enumerable: true,
                    get: function get() {
                        return o.creepName;
                    }
                },
                spawning: {
                    enumerable: true,
                    get: function get() {
                        return false;
                    }
                },
                my: {
                    enumerable: true,
                    get: function get() {
                        return o.user == runtimeData.user._id;
                    }
                },
                body: {
                    enumerable: true,
                    get: function get() {
                        return _.map(o.creepBody, function (type) {
                            return { type: type, hits: 0 };
                        });
                    }
                },
                owner: {
                    enumerable: true,
                    get: function get() {
                        return _.isUndefined(o.user) || o.user === null ? undefined : {
                            username: runtimeData.users[o.user].username
                        };
                    }
                },
                ticksToLive: {
                    enumerable: true,
                    get: function get() {
                        return o.creepTicksToLive;
                    }
                },
                carryCapacity: {
                    enumerable: true,
                    get: function get() {
                        return _.reduce(o.creepBody, function (result, type) {
                            return result += type === C.CARRY ? C.CARRY_CAPACITY : 0;
                        }, 0);
                    }
                },
                carry: {
                    enumerable: true,
                    get: function get() {
                        return { energy: 0 };
                    }
                },
                fatigue: {
                    enumerable: true,
                    get: function get() {
                        return 0;
                    }
                },
                hits: {
                    enumerable: true,
                    get: function get() {
                        return 0;
                    }
                },
                hitsMax: {
                    enumerable: true,
                    get: function get() {
                        return o.creepBody.length * 100;
                    }
                },
                saying: {
                    enumerable: true,
                    get: function get() {
                        return o.creepSaying;
                    }
                }
            });
            return creep;
        }
    });

    Tombstone.prototype.toString = register.wrapFn(function () {
        return '[Tombstone #' + this.id + ']';
    });

    Object.defineProperty(globals, 'Tombstone', { enumerable: true, value: Tombstone });
};
